src/traces/waterfall/hover.js

```diff
@@ -22,28 +23,35 @@ module.exports = function hoverPoints(pointData, xval, yval, hovermode) {
 
     var cd = point.cd;
     var trace = cd[0].trace;
+    var isHorizontal = (trace.orientation === 'h');
+
+    var vAxis = isHorizontal ? pointData.xa : pointData.ya;
+
+    function formatNumber(a) {
+        return hoverLabelText(vAxis, a);
+    }
 
     // the closest data point
     var index = point.index;
     var di = cd[index];
 
-    var sizeLetter = (trace.orientation === 'h') ? 'x' : 'y';
+    var sizeLetter = isHorizontal ? 'x' : 'y';
 
     var size = (di.isSum) ? di.b + di.s : di.rawS;
 
     if(!di.isSum) {
         // format delta numbers:
         if(size > 0) {
-            point.extraText = size + ' ' + DIRSYMBOL.increasing;
+            point.extraText = formatNumber(size) + ' ' + DIRSYMBOL.increasing;
         } else if(size < 0) {
-            point.extraText = '(' + (-size) + ') ' + DIRSYMBOL.decreasing;
+            point.extraText = '(' + (formatNumber(-size)) + ') ' + DIRSYMBOL.decreasing;
         } else {
             return;
         }
         // display initial value
-        point.extraText += '<br>Initial: ' + (di.b + di.s - size);
+        point.extraText += '<br>Initial: ' + formatNumber(di.b + di.s - size);
     } else {
-        point[sizeLetter + 'LabelVal'] = size;
+        point[sizeLetter + 'LabelVal'] = formatNumber(size);
```

Comment:
not quite sure where this would break down, but in principle `*LabelVal` should be a number, and `*Label` should be the string you want to use to display that number.

Reply:
Perhaps non-default separators, or tickprefix/suffix, would break this as is

Reply:
In `bar` traces we use a similar pattern: https://github.com/plotly/plotly.js/blob/07b6c9c296b2037ebda7d12619e31354fc263296/src/traces/bar/hover.js#L138.
Do you think there would be a problem there?

Reply:
Bar uses `*LabelVal` correctly, it's a number. I'm guessing that if there's no `*Label` then later on `*LabelVal` gets converted to a string using `hoverLabelText` but I don't know. If so, perhaps you can just leave this one as a number too.

Whew, the hover system is certainly ripe for refactoring... or at least documenting.

Reply:
> Whew, the hover system is certainly ripe for refactoring... or at least documenting.

:pray:

Reply:
> Bar uses `*LabelVal` correctly, it's a number. I'm guessing that if there's no `*Label` then later on `*LabelVal` gets converted to a string using `hoverLabelText` but I don't know. If so, perhaps you can just leave this one as a number too.

OK. Let's keep this as it is in this PR.